app/modules/cron.c

```diff
@@ -0,0 +1,223 @@
+  cronent_ud_t *eud;
+  size_t i;
+  for (i = 0; i < cronent_count; i++) {
+    lua_rawgeti(L, LUA_REGISTRYINDEX, cronent_list[i]);
+    eud = lua_touserdata(L, 1);
+    if (eud == ud) break;
+  }
+  if (i == cronent_count) return -1;
+  return i;
+}
+
+static int lcron_schedule(lua_State *L) {
+  cronent_ud_t *ud = luaL_checkudata(L, 1, "cron.entry");
+  char *strdesc = (char*)luaL_checkstring(L, 2);
+  struct cronent_desc desc;
+  lcron_parsedesc(L, strdesc, &desc);
+  ud->desc = desc;
+  size_t i = lcron_findindex(L, ud);
+  if (i == -1) {
+    lua_pushvalue(L, 1);
+    cronent_list = os_realloc(cronent_list, sizeof(int) * (cronent_count + 1));
+    cronent_list[cronent_count++] = lua_ref(L, LUA_REGISTRYINDEX);
+  }
+  return 0;
+}
+
+static int lcron_handler(lua_State *L) {
+  cronent_ud_t *ud = luaL_checkudata(L, 1, "cron.entry");
+  luaL_checkanyfunction(L, 2);
+  lua_pushvalue(L, 2);
+  luaL_unref(L, LUA_REGISTRYINDEX, ud->cb_ref);
+  ud->cb_ref = luaL_ref(L, LUA_REGISTRYINDEX);
+  return 0;
+}
+
+static int lcron_unschedule(lua_State *L) {
+  cronent_ud_t *ud = luaL_checkudata(L, 1, "cron.entry");
+  size_t i = lcron_findindex(L, ud);
+  if (i == -1) return 0;
+  luaL_unref(L, LUA_REGISTRYINDEX, cronent_list[i]);
+  memmove(cronent_list + i, cronent_list + i - 1, sizeof(int) * cronent_count - i - 1);
+  cronent_count--;
+  return 0;
+}
+
+static int lcron_delete(lua_State *L) {
+  cronent_ud_t *ud = luaL_checkudata(L, 1, "cron.entry");
+  lcron_unschedule(L);
+  luaL_unref(L, LUA_REGISTRYINDEX, ud->cb_ref);
+  return 0;
+}
+
+static void cron_handle_time(uint8_t mon, uint8_t dom, uint8_t dow, uint8_t hour, uint8_t min) {
+  lua_State *L = lua_getstate();
+  struct cronent_desc desc;
+  desc.mon  = (uint16_t)1 << (mon - 1);
+  desc.dom  = (uint32_t)1 << (dom - 1);
+  desc.dow  = ( uint8_t)1 << dow;
+  desc.hour = (uint32_t)1 << hour;
+  desc.min  = (uint64_t)1 << min;
+  for (size_t i = 0; i < cronent_count; i++) {
+    lua_rawgeti(L, LUA_REGISTRYINDEX, cronent_list[i]);
+    cronent_ud_t *ent = lua_touserdata(L, 1);
+    lua_pop(L, 1);
+    if ((ent->desc.mon  & desc.mon ) == 0) continue;
+    if ((ent->desc.dom  & desc.dom ) == 0) continue;
+    if ((ent->desc.dow  & desc.dow ) == 0) continue;
+    if ((ent->desc.hour & desc.hour) == 0) continue;
+    if ((ent->desc.min  & desc.min ) == 0) continue;
+    lua_rawgeti(L, LUA_REGISTRYINDEX, ent->cb_ref);
+    lua_rawgeti(L, LUA_REGISTRYINDEX, cronent_list[i]);
+    lua_call(L, 1, 0);
+  }
+}
+
+static void cron_handle_tmr() {
+  struct rtc_timeval tv;
+  rtc_time_gettimeofday(&tv);
+  if (tv.tv_sec == 0) { // Wait for RTC time
+    ets_timer_arm_new(&cron_timer, 1000, 0, 1);
+    return;
+  }
+  time_t t = tv.tv_sec;
+  struct tm tm;
+  gmtime_r(&t, &tm);
+  uint32_t diff = 60000 - tm.tm_sec * 1000 - tv.tv_usec / 1000;
+  if (tm.tm_sec == 59) {
+    t++;
+    diff += 60000;
+    gmtime_r(&t, &tm);
+  }
+  if (tm.tm_sec <= 1) {
```

Comment:
I'm not convinced that you want this check here. If the timer is delayed by 10 seconds (due to other activity), you still want to run the scheduled tasks.